### PR TITLE
Initialization design doc

### PR DESCRIPTION
**Pull request by tonyyang-svail**
Conclusion: two separate programDesc.

### CODE REVIEW

**Comment by helinwang** [comment]
Option 1 (Two separate ProgramDesc): we need somehow separate the ProgramDesc, the separation process is already similar to prune, maybe we can just do prune. So probably this is not a good option.
Option 2 (run once): In my opinion we should not do this, since it's possible that the user want to initialize the network during training.
Option 3 (Remove Init operator during Prune):
1. we should not remove init OP if the targets contains the init OP.
1. In my opinion we need to analyze the dependency correctly to resolve what needs to be run (our dependency graph should not have cycles), not just do regular expression.

**Comment by wangkuiyi** [comment]
https://travis-ci.org/PaddlePaddle/Paddle/jobs/288826972#L993-L998

**Comment by wangkuiyi** [comment]
I vote for Proposal 1. -- make it run once.

**Comment by tonyyang-svail** [comment]
After offline discussion with @reyoung, the current implementation will be based on proposal 1.

File: doc/design/initialization.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+PaddlePaddle needs a way to init parameters.
+
+## Challenge
+
+Initialization operators must run once and only one; otherwise, every iteration would clear the parameters.
+
+## Proposals
+
+### Proposal 1: Two seperate `ProgramDesc`.
```
Comment (Superjom) [inline doc/design/initialization.md]: this seems simple and workable.

The resource which survived across multiple block/program is the `Variables` in `Scopes`.

A program is the container of multiple Blocks, more than one program seems not a problem.

**Comment by jacquesqiao** [comment]
if we have a conclusion that we will use multiple programs, one is for initializing, should we update this design?

File: doc/design/initialization.md [hunk 2]
```diff
@@ -8,41 +8,6 @@ PaddlePaddle needs a way to init parameters.
 
 Initialization operators must run once and only one; otherwise, every iteration would clear the parameters.
 
-## Proposals
-
-### Proposal 1: Two seperate `ProgramDesc`.
+## Solution: Two seperate `ProgramDesc`.
```
Comment (helinwang) [inline doc/design/initialization.md]: Who will output the two separate `ProgramDesc`s?

I my mind:

Python ---(entire `ProgramDesc`) ---> Pruner ---(ProgramDesc1, ProgramDesc)---> Executer.

We need to make sure Pruner have the global view. For example, in cluster training, the Pruner need to have the global view in order to do things like deciding parameter placement.

Comment (helinwang) [inline doc/design/initialization.md]: seperate -> separate

Comment (tonyyang-svail) [reply doc/design/initialization.md]: Python generates two separate `ProgramDesc`, one for initialization, and one for training. Both `ProgramDesc`s contain all the parameter description. Do you think it is enough to decide parameter placement? If not, we can always send both `ProgramDesc`s to Pruner.

Comment (helinwang) [reply doc/design/initialization.md]: If Python can do the separation, pruner should be able to do it as well. Is it possible to let pruner do the job?

Parameter placement is just an example supporting the argument that the pruner needs to have the global view.

**Pull request closed without merging**